pgmtoasciiart.1
.TH PGMTOASCIIART 1 "14 November 2010"
.SH NAME
pgmtoasciiart \- convert PGM to ASCII art
.SH SYNOPSIS
.B pgmtoasciiart
.RI [ options... ]
.RI [ filename ]
.SH DESCRIPTION
.PP
\fBpgmtoasciiart\fP takes \fInetpbm\fP PGM format and converts
it to "ASCII art" by picking an ASCII character with
appropriate black density for each
pixel of the input image.
.PP
This is a low-level tool.  You may prefer to do this
conversion with the help of
.BR mkasciiart (1).
.SH USAGE
.PP
Options are as follows:
.TP
.B "-reverse"
Reverses the tonal scale, producing an image that is
intended to be displayed as white-on-black.
.TP
.B "-clamp"
Rather than compress the contrast due to the non-blackness
of the darkest character, just give all input pixels darker
than the darkest character the darkest character value.
.TP
.B "-threshold"
When the the input pixel lies between two characters'
values, rather than choose an output character
weighted-randomly, just pick the closest.
.TP
.BI "-scale " string
The given \fIstring\fP is treated as a list of ASCII
characters by increasing black density.  The output gray
level for an input character will be determined by the
position of the character in this list.  Characters not in
the list will be treated as black.  The space character is
always white, and should be omitted.
.IP
When neither \fB-scale\fP or \fB-tonescale\fP are given, the
output gray level will be determined directly by the
measured proportional density of the input character. See
the \fB-font\fP flag to select which font information will
be used in this calculation.
.TP
.B "-tonescale"
Like \fB-scale\fP, except that a default scale will be
constructed by sorting the glyphs according to character
density taken from font information.
.TP
.BI "-font " tag
The given font \fItag\fP is used to determine which built-in scale to
use. Different fonts have different character densities, so this
argument may be useful in dealing with ASCII art that was
produced with a particular font in mind.  Available
font tags as of this writing are "sans", which corresponds
to \fIBitstream Vera Sans Mono\fP, and "typewriter", which
corresponds to \fICourier New\fP.
.SH NOTES
.PP
This is a pretty simplistic implementation, and could be
improved quite a lot.
.PP
The font tones have been measured automatically by rendering
characters. Character patterning is in no way taken into
account in this process.
.SH AUTHOR
Bart Massey <bart@cs.pdx.edu>
.SH "SEE ALSO"
.BR mkasciiart (1),
.BR asciiarttopgm (1),
.BR asciiarttopbm (1),
.BR netpbm (1),
.BR pgm (5)


pgmtoasciiart.c
/*
 * Copyright © 2010 Bart Massey
 * ALL RIGHTS RESERVED
 * [This program is licensed under the "MIT License"]
 * Please see the file COPYING in the source
 * distribution of this software for license terms.
 */

#include <assert.h>
#include <pgm.h>
#ifdef DEBIAN
#define optParseOptions3 pm_optParseOptions3
#else
#include <shhopt.h>
#endif
#include <stdio.h>
#include <string.h>

#include "glyphshades.h"

static char *usage = "usage: pgmtoasciiart [-reverse] [-clamp] [-threshold]\n    [-f <font-tag>] [-tonescale] [-scale <chars>] [file]";

static int maxscale;      /* length of scale (i.e. gray-levels available) */
static int mingray = 1;   /* if it aint at least this, it's white */
static struct glyphshade *shade;    /* shade information */ 
static int gscale;        /* adjustment for very HDR inputs */

/* command-line arguments */
static FILE *fp;
static char *scale;   /* for pointing to the gray-level being used */
static char *font_tag;
static unsigned int tonescale, reverse, clamp, threshold;

static void parse_command_line(int argc, char ** argv) {
    optEntry *option_def = malloc(100 * sizeof(optStruct));
    optStruct3 opt;
    unsigned int option_def_index;
    int i;

    option_def_index = 0;
    OPTENT3(0, "scale", OPT_STRING, &scale, NULL, 0);
    OPTENT3(0, "tonescale", OPT_FLAG, NULL, &tonescale, 0);
    OPTENT3(0, "reverse", OPT_FLAG, NULL, &reverse, 0);
    OPTENT3(0, "clamp", OPT_FLAG, NULL, &clamp, 0);
    OPTENT3(0, "threshold", OPT_FLAG, NULL, &threshold, 0);
    OPTENT3(0, "font", OPT_STRING, &font_tag, NULL, 0);

    reverse = 0;
    clamp = 0;
    threshold = 0;
    fp = stdin;
    scale = 0;
    tonescale = 0;
    font_tag = "sans";
    shade = 0;

    opt.opt_table = option_def;
    opt.short_allowed = 0;  /* We have no short (old-fashioned) options */
    opt.allowNegNum = 1;  /* We may have parms that are negative numbers */

    optParseOptions3(&argc, argv, opt, sizeof(opt), 0);

    if (argc > 2)
        pm_error(usage);
    for (i = 0; shades[i].font_tag; i++) {
        if (!strcmp(font_tag, shades[i].font_tag)) {
            if (tonescale && !scale)
                scale = *shades[i].scale;
            shade = shades[i].shades;
            break;
        }
    }
    if (!shade)
        pm_error("unknown font tag %s", font_tag);
    if (argc == 2) {
        fp = fopen(argv[1], "r");
        if (!fp)
            pm_error("could not open input file");
    }
}

static char shadechar(unsigned int g, unsigned int nscale) {
    float w, frac;
    int lower, upper;
    
    g >>= gscale;
    w = g / (float) nscale;
    if (!reverse)
        w = 1.0 - w;
    if (!clamp)
        w *= shade[94].w;
    if (w >= shade[94].w)
        return shade[94].ch;
    lower = 0;
    upper = 94;
    /* XXX This could be replaced by a fast binary search,
       but how fast can it even need to be? */
    while (shade[lower + 1].w <= w)
        lower++;
    while (shade[upper - 1].w >= w)
        upper--;
    if (lower == upper)
        return shade[lower].ch;
    frac = (w - shade[lower].w) / (shade[upper].w - shade[lower].w);
    if (threshold) {
        if (frac >= 0.5)
            return shade[upper].ch;
        return shade[lower].ch;
    }
    if (random() % 100 >= 100 * frac)
        return shade[upper].ch;
    return shade[lower].ch;
}

static char scalechar(unsigned int g, unsigned int nscale) {
    int k = (maxscale - 1) * (nscale - g) / nscale;

    assert(k >= 0 && k < maxscale);
    if (reverse) {
        if (k >= maxscale - mingray)
            return ' ';
        return scale[maxscale - mingray - k];
    }
    if (k < mingray)
        return ' ';
    return scale[k - mingray];
}


static void onechar(unsigned int g, unsigned int nscale) {
    if (scale)
        putchar(scalechar(g, nscale));
    else
        putchar(shadechar(g, nscale));
}

int main(int argc, char **argv) {
    int cols, rows;
    unsigned int nscale;
    gray **gp;
    int i, j;

    pgm_init(&argc, argv);
    parse_command_line(argc, argv);

    if (scale)
        maxscale = strlen(scale) + mingray;
    gp = pgm_readpgm(fp, &cols, &rows, &nscale);
    gscale = 0;
    while (nscale >= (1 << 16)) {
        nscale >>= 1;
        gscale++;
    }
    for (i = 0; i < rows; i++) {
        for (j = 0; j < cols; j++)
            onechar(gp[i][j], nscale);
        putchar('\n');
    }
    return 0;
}
